application-xpoll-api/src/main/java/com/xwiki/xpoll/internal/DefaultXPollManager.java

```
@@ -0,0 +1,209 @@
+                documentReference), e);
+        }
+    }
+
+    @Override
+    public Map<String, Integer> getVoteResults(DocumentReference documentReference)
+    {
+        XWikiContext context = contextProvider.get();
+        try {
+            XWikiDocument doc = context.getWiki().getDocument(documentReference, context);
+            List<BaseObject> xpollVotes = doc.getXObjects(XPOLL_VOTES_CLASS_REFERENCE);
+            BaseObject xpollObj = doc.getXObject(XPOLL_CLASS_REFERENCE);
+            List<String> proposals = xpollObj.getListValue(PROPOSALS);
+            return getXPollResults(xpollVotes, proposals);
+        } catch (XWikiException e) {
+            logger.warn("Failed to retrieve the vote results for poll [{}]. Root cause: [{}].", documentReference,
+                ExceptionUtils.getRootCauseMessage(e));
+            return new HashMap<>();
+        }
+    }
+
+    private void setUserVotes(List<String> votedProposals, XWikiContext context, XWikiDocument doc,
+        DocumentReference user) throws XWikiException
+    {
+        String currentUserName = serializer.serialize(user, doc.getDocumentReference().getWikiReference().getName());
+        BaseObject xpollVoteOfCUrrentUser = doc.getXObject(XPOLL_VOTES_CLASS_REFERENCE, USER,
+            currentUserName, false);
+
+        if (xpollVoteOfCUrrentUser == null) {
+            xpollVoteOfCUrrentUser = doc.newXObject(XPOLL_VOTES_CLASS_REFERENCE, context);
+        }
+
+        xpollVoteOfCUrrentUser.set(USER, currentUserName, context);
+        xpollVoteOfCUrrentUser.set(VOTES, votedProposals, context);
+    }
+
+    private void updateWinner(XWikiContext context, XWikiDocument doc) throws XWikiException
+    {
+        BaseObject xpollObj = doc.getXObject(XPOLL_CLASS_REFERENCE);
+        List<BaseObject> xpollVotes = doc.getXObjects(XPOLL_VOTES_CLASS_REFERENCE);
+
+        List<String> proposals = xpollObj.getListValue(PROPOSALS);
```

This can throw a ``NullPointerException``. See my previous comment. Would be better to throw an ``XPollException`` if the ``XPollClass`` object is missing.